supplementary-guidelines/semantic_conventions_code_generation.md

```diff
@@ -0,0 +1,118 @@
+# Semantic convention libraries
+
+<!-- toc -->
+
+- [Stability and Versioning](#stability-and-versioning)
+  - [Deprecated Conventions](#deprecated-conventions)
+- [Semantic Conventions Artifact Structure](#semantic-conventions-artifact-structure)
+- [Generating semantic conventions](#generating-semantic-conventions)
+
+<!-- tocstop -->
+
+The code for OpenTelemetry Semantic Conventions defined in this repository can be auto-generated.
+
+OpenTelemetry Language SIGs can generate Semantic Conventions code in the form that's idiomatic for
+their language and may (or may not) ship it as a stand-alone library.
+
+This document outlines common patterns and provides non-normative guidance on how to structure semantic conventions artifact
+and generate the code.
+
+## Stability and Versioning
+
+Semantic Conventions contain a mix of stability levels.
+Language SIGs that ship semantic conventions library may decide to ship a stable artifact with stable part of the Semantic Conventions, an experimental artifact with all Semantic Conventions, or other combination that's idiomatic for this language and provides [SemVer 2.0](https://semver.org/) stability guarantees.
+
+Possible solutions include:
+
+- Generate all Semantic Conventions for a given version in specific folder while keeping old versions intact. It is used by [opentelemetry-go](https://github.com/open-telemetry/opentelemetry-go/tree/main/semconv/) but could be problematic if the artifact size is a concern.
+- Follow language-specific conventions to annotate experimental parts. For example, Semantic Conventions in Python puts experimental attributes in `opentelemetry.semconv._incubating` import path which is considered (following Python underscore convention) to be internal and subject to change.
+- Ship two different artifacts: one that contains stable Semantic Conventions and another one that contains all conventions.
+
+> Note:
+> Shipping two versions of the same artifact (stable and preview) could be problematic due to diamond-dependency problems.
+> For example, if user application depends on the `semconv v1.0.0-preview` and some library brings transitive dependency on `semconv v1.1.0` that does not contain
+> experimental convention, the latter would be resolved leading to compilation or runtime issues in the application.
+
+Instrumentation libraries should depend on the stable (parts of) semantic convention artifact or copy relevant definitions into their own code base.
+Incubating parts are intended for the end-users.
```

Comment:
What does "Incubating parts" mean here?

Reply:
We talk about Incubating in several places it seems it's never defined what it means. above we kinda say it's where experimental attributes should go in, but here we say they are intended for end user. Below we say it's for stable conventions. I'm  a bit confused 🤔

Reply:
I've updated the text to avoid using it except in the naming suggestions.

It effectively means either a package like `semconv-incubating` that contains all the semantic conventions (experimental and stable). Or, if there is just one package, it's a part that's annotated as not stable. E.g. in python it's `semconv._incubating`.

I now use `preview` to refer to that artifact, lmk if it's better or if there are parts I still need to clarify. Thanks!